File: src/fibonacci.c
#include "fibonacci.h"

int fibonacci(const int n) {
  if (n == 0 || n == 1) {
    return n;
  }
  return fibonacci(n - 1) + fibonacci(n - 2);
}

int fibonacci_array(const int n) {
  int array[n];
  array[0] = 0;
  array[1] = 1;

  for (int i = 2; i <= n; i++) {
    array[i] = array[i - 1] + array[i - 2];
  }
  return array[n];
}

int dynamic_fibonacci(const int n) {
  int x = 0;
  int y = 1;
  for (int i = 0; i < n; i++) {
    const int t = y;
    y += x;
    x = t;
  }
  return x;
}

File: src/fibonacci.rs
use std::vec;

pub fn recursive_fibonacci(n: usize) -> usize {
    if n == 0 || n == 1 {
        return n;
    }
    recursive_fibonacci(n - 1) + recursive_fibonacci(n - 2)
}

pub fn array_fibonacci(n: usize) -> usize {
    let mut vector = vec![0; n + 1];
    vector[0] = 0;
    vector[1] = 1;

    for i in 2..=n {
        let x = vector[i - 1];
        let y = vector[i - 2];
        vector[i] = x + y;
    }

    vector[n]
}

pub fn dynamic_fibonacci(n: usize) -> usize {
    let mut x = 0;
    let mut y = 1;

    for _i in 0..n {
        let t = y;
        y += x;
        x = t;
    }

    x
}
